File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.java
import java.util.*;
import java.io.*;

public class Main {
    static class Node {
        int cur;
        int time;

        public Node(int cur, int time) {
            this.cur = cur;
            this.time = time;
        }

    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int cnt = 0;
        int[] visited = new int[100001];
        Arrays.fill(visited, 100001);
        int shortest = 100001;

        if (n >= m) {
            System.out.println(n-m);
            System.out.println(1);
            System.exit(0);
        }

        LinkedList<Node> que = new LinkedList<>();
        que.add(new Node(n, 0));

        while (que.size() > 0) {
            Node node = que.pollFirst();

            if (node.time > shortest) {
                break;
            }

            if (node.cur == m) {
                shortest = node.time;
                cnt ++;
                continue;
            }

            if (visited[node.cur] < node.time) {
                continue;
            } else {
                visited[node.cur] = node.time;
            }

            for (int v : new int[] {node.cur-1, node.cur+1, node.cur*2}) {
                if (0 <= v && v <= 100000 && visited[v] >= node.time+1) {
                    que.add(new Node(v, node.time+1));
                }
            }
        }

        System.out.println(shortest);
        System.out.println(cnt);
    }
}

File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.py
# 숨바꼭질2
import sys
input = sys.stdin.readline
from collections import deque

n, m = map(int,input().split())
shortest = 100000
cnt = 0
visited = [100001] * 100001

if n >= m:
    print(n - m)
    print(1)
    exit()

que = deque([])
que.append((n, 0))

while que:
    cur, time = que.popleft()

    if time > shortest:
        break
    
    if cur == m:
        shortest = time
        cnt += 1
        continue

    if visited[cur] < time:
        continue
    else:
        visited[cur] = time

    for v in [cur-1, cur+1, cur*2]:
        if 0 <= v <= 100000 and visited[v] >= time+1:
            que.append((v, time+1))

print(shortest)
print(cnt)
